Kestrel SNI from config

halter73 opened the pull request:
```
{
  "Kestrel": {
    "Endpoints": {
      "EndpointName": {
        "Url": "https://*",
        "Sni": {
          "a.example.org": {
            "Protocols": "Http1AndHttp2",
            "SslProtocols": [ "Tls11", "Tls12", "Tls13"],
            "Certificate": {
              "Path": "testCert.pfx",
              "Password": "testPassword"
            },
            "ClientCertificateMode" : "NoCertificate"
          },
          "*.example.org": {
            "Certificate": {
              "Path": "testCert2.pfx",
              "Password": "testPassword"
            }
          },
          "*": {
            // At least one subproperty needs to exist per SNI section or it
            // cannot be discovered via IConfiguration
            "Protocols": "Http1",
          }
        }
      }
    }
  }
}
```


- What is this feature?
  - Adding support for configuring [Kestrel's endpoint-specific options](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-3.1#endpoint-configuration), most notably certs, via config (e.g. appsettings.json). The complete list of configurable options are currently as follows:
    - Certificate
    - ClientCertficateMode
    - SslProtocols
    - HttpProtocols

- What is supported?
  - Server name (or host name)
    - Exact match (StringComparision.OrdinalIgnoreCase)
      - Exact matches will always be selected before any wildcard matches
    - Wildcard prefix
      - If the server name specified in config starts with "*.", and the requested server name ends with the string that fallows (again StringComparision.OrdinalIgnoreCase), that's considered a match. The longest match will be taken in the event there are 
    - Full wildcard
      - If specified, any requested server name that doesn't have an exact match or wildcard prefix match will match this. This will be matched even if the client doesn't request any server name (i.e. doesn't use SNI).

- Where/How it can fail?
  - If the client requests a server name via SNI that doesn't match any part of SNI config (this means that no "full wildcard" config was specified), the connection will be refused.
  - If the client requests no server name (i.e. doesn't use SNI) and no "full wildcard" config was specified, the connection will be refused.
  - Certs for all configurations are now loaded at startup. If a cert isn't found or fails to load for a specific SNI endpoint, the server will fail to start like it does today for an endpoint configured without SNI.
  - If a cert isn't found or fails to load after configuration changes at runtime, the server will log critically like it does today for an endpoint configured without SNI.

- What APIs are exposed (if any)?
  - No APIs are exposed yet. Something similar to `internal static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOptionsSelectionCallback serverOptionsCallback, object state = null)` could be made public once we figure out some kinks like how the handshake timeout can be configured with that overload.
    - Do we want a new options object that has only a few properties like a ServerOptionsSelectionCallback and a HandshakeTimeout? We could add convenience methods that make it easy to set common configuration like the server certificate without needing to manually replace or wrap the ServerOptionsSelectionCallback. The convenience methods would likely wrap the ServerOptionsSelectionCallback internally. A downside to this approach is things like a server certificate getter would be difficult or impossible.
    - Do we want to add an additional parameter for the handshake timeout? This risks creating too many UseHttps() overloads if there wind up being too many other things that cannot be configured via the ServerOptionsSelectionCallback.
    - Do we want to expand the HttpsConnectionAdapterOptions used by the existing UseHttps() method? This avoids adding a new overload to UseHttps(), but it makes HttpsConnectionAdapterOptions more complicated and confusing that it already does. It becomes unclear if a ServerOptionsSelectionCallback overrides OnAuthenticate, ServerCertificate, etc...

- How it composes with anything we already have built in? (What the fallbacks are)
    - The SNI section only affects HTTPS endpoints.
    - Configuration specified directly in the SNI section overrides all other configuration when that SNI section is matched.
    - If an Endpoint does have its own SNI section, that will take precedents over the "EndpointDefaults" SNI section. No merging of the two SNI sections will occur.
    - ~If an Endpoint doesn't have its own SNI section, but an "EndpointDefaults" SNI section does exist, the configuration logic behaves as if the SNI section was defined directly on the endpoint. This means that a "Certificate" defined directly on the endpoint could be overridden by a certificate defined in an "EndpointDefaults" SNI section.~
    - If no relevant configuration exists in the selected SNI section for a given property (e.g. Certificate, ClientCertficateMode, SslProtocols, HttpProtocols), named-endpoint-wide configuration will be used instead.
      - Named-endpoint-wide configuration can be overridden as always via `public KestrelConfigurationLoader Endpoint(string name, Action<EndpointConfiguration> configureOptions)`. Since there's no public API currently exposing SNI-specific config yet, only named-endpoint-wide configuration, non SNI-specific configuration" can be overridden this way.
    - If no relevant configuration exists in the selected SNI or named endpoint section for a given property, ~"EndpointDefaults" and/or~ "Certificates:Default/Development" configuration will be used instead.
    - If no relevant configuration exists in the selected SNI or named endpoint section for a given property, and no defaults are configured via ~"EndpointDefaults" and/or~ "Certificates:Default/Development", configuration will fallback to what was configured via `KSO.ConfigureHttpsDefaults(Action<HttpsConnectionAdapterOptions> configureOptions)` or `KSO.ConfigureEndpointDefaults(Action<ListenOptions> configureOptions)`.

- We decided **against** adding an SNI section to "EndpointDefaults", because certificates defined in "EndpointDefaults:SNI" it had the weird behavior of overriding "Certificate"s defined directly on endpoints or via `ListenOptions.UseHttps()` since certs defined via SNI config is preferred over normal certs.

```
    // The following was considered but decided against.
    //
    // EndpointDefaults will only be used only for endpoints without their own SNI configuration.
    // SNI configuration will always be preferred even over endpoint-wide non-SNI endpoint-specific
    // Configuration, but if no equivalent setting exists in the SNI configuration for
    // the given property, it will fallback to endpoint specific configuration, then code-configured endpoint/HTTPS defualts,
    // then non-SNI configuration EndpointDefaults.
    "EndpointDefaults": {
      "Sni": {
        "a.example.org": {
          "Certificate": {
            "Path": "testCert3.pfx",
            "Password": "testPassword"
          },
        },
        "b.example.org": {
          "Certificate": {
            "Path": "testCert4.pfx",
            "Password": "testPassword"
          }
        }
      }
    }
```

Addresses #15144

halter73 committed:
Kestrel SNI from config

halter73 committed:
ConfigurationReaderTests

jkotalik commented:
Glad I didn't bet that you would get this done.

jkotalik commented on src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs:
```diff
@@ -196,14 +260,60 @@ public IConfigurationSection ConfigSection
+        }
+    }
+
+    internal class SniConfig
```
nit: private class or put in separate file.

jkotalik replied:
Nvrm, I guess all of these are in the same class.

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs:
```diff
@@ -97,7 +98,8 @@ private IEnumerable<EndpointConfig> ReadEndpoints()
                     SslProtocols = ParseSslProcotols(endpointConfig.GetSection(SslProtocolsKey)),
-                    ClientCertificateMode = ParseClientCertificateMode(endpointConfig[ClientCertificateModeKey])
+                    ClientCertificateMode = ParseClientCertificateMode(endpointConfig[ClientCertificateModeKey]),
+                    SNI = ReadSni(endpointConfig.GetSection(SNIKey)),
```
Consider a global SNI section as well that takes priority over the default cert, but is overridden by endpoint certs.

Tratcher commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -346,6 +367,82 @@ public void Load()
+
+            if (!string.IsNullOrEmpty(serverName))
+            {
+                foreach (var (name, sniConfigCandidate) in endpoint.SNI)
```
Pre-build a `Dictionary<string, SslServerAuthenticationOptions>` ignore case collection when loading config. Consider setting a flag that says if the SNI contains any top level wildcards (`*.domain`) or a global wildcard (`*`). That way you can avoid enumerating the dictionary if there aren't any.

Matching:
1. If there's no serverName check the dictionary for a top level wildcard `*`. If not present then fail.
1. Direct lookup in the dictionary (fastest)
1. Scan the dictionary for the most specific wildcard `*.domain`, or top level wildcard `*`.
1. Fail if there's no match or wildcard in the dictionary.

Tratcher commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -346,6 +367,82 @@ public void Load()
+                    }
+                    // Note that we only slice off the `*`. We want to match the leading `.` also.
+                    else if (name.StartsWith(wildcardPrefix, StringComparison.Ordinal)
+                        && serverName.EndsWith(name.Substring(wildcardHost.Length), StringComparison.OrdinalIgnoreCase)
```
Substring? Is there a better Span or StringSegment version?

Tratcher commented on src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:
```diff
@@ -227,5 +228,26 @@ public static ListenOptions UseHttps(this ListenOptions listenOptions, HttpsConn
+        /// <param name="serverOptionsCallback">Callback to configure HTTPS options.</param>
+        /// <param name="state">State for the <see cref="ServerOptionsSelectionCallback" />.</param>
+        /// <returns>The <see cref="ListenOptions"/>.</returns>
+        internal static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOptionsSelectionCallback serverOptionsCallback, object state = null)
```
Consider making this public. We want the config to use a consistent API with whatever the eventual public API for this callback is.

davidfowl reviewed (changes requested):
@halter73 I'd like a full description of the behavior in the PR description:
- What is this feature?
- What is supported?
  - Wildcards  (how matching works)
  - Explicit host names
  - etc.
- Where/How it can fail?
- What APIs are exposed (if any)?
- How it composes with anything we already have built in?
- What the fallbacks are (I saw a discussion about falling back to the default certificate in some cases).

halter73 committed:
Cache SslServerAuthenticationOptions and certs

halter73 commented on src/Servers/Kestrel/Core/src/Internal/SniOptionsSelector.cs:
```diff
@@ -0,0 +1,116 @@
+                }
+                else
+                {
+                    throw new AuthenticationException(CoreStrings.FormatSniNotConfiguredForServerName(serverName, _endpointName));
```
@Tratcher One unfortunate thing about this, is that you can't use an empty wildcard SNI section (e.g. `"*": {}`) to get endpoint-default behavior for non-matches, because IConfigurationSection.GetChildren() doesn't return empty children. You need at least one sub-property 😦

Tratcher replied:
Should it be an array instead? Then that one property would be the name.

halter73 committed:
Validate cached options on load

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs:
```diff
@@ -106,6 +108,46 @@ private IEnumerable<EndpointConfig> ReadEndpoints()
+            {
+                // "SNI": {
+                //     "a.example.org": {
+                //         "Protocols": "Http1AndHttp2",
```
This protocol is used for ALPN. How does it interact with the protocol configured on the endpoint? Do we need to flow this value up in the handshake results feature so we know what was requested vs what was negotiated?

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs:
```diff
@@ -20,6 +20,7 @@ internal class ConfigurationReader
         private const string EndpointsKey = "Endpoints";
         private const string UrlKey = "Url";
         private const string ClientCertificateModeKey = "ClientCertificateMode";
+        private const string SNIKey = "SNI";
```
bikeshedding: Sni? Naming guidelines is acronym of 3 or more letters are camelcased, e.g. Url, Ssl, Tls.

halter73 committed:
WIP

BrennanConroy commented on src/Servers/Kestrel/Core/src/CoreStrings.resx:
```diff
@@ -620,4 +620,10 @@ For more information on configuring HTTPS see https://go.microsoft.com/fwlink/?l
+    <value>Connection refused because no SNI configuration was found for '{serverName}' in '{endpointName}'. To allow all connections, add a wildcard ('*') SNI section.</value>
+  </data>
+  <data name="SniNotConfiguredToAllowNoServerName" xml:space="preserve">
+    <value>Connection refulsed because the client did not specify a server name, and no wildcard ('*') SNI configuration was found in '{endpointName}'.</value>
```
New word **refulsed**
```suggestion
    <value>Connection refused because the client did not specify a server name, and no wildcard ('*') SNI configuration was found in '{endpointName}'.</value>
```

halter73 committed:
Support default handshake timeout

halter73 committed:
Flow SNI HttpProtocols to middleware

halter73 committed:
Use default CertificateRevocationCheckMode for SNI endpoints

halter73 committed:
Solidify fallback logic

Tratcher reviewed (commented):
Only minor feedback at the moment.

The big items that still need to be addressed:
- Should there be a global SNI config? E.g. so you can have 10 endpoints all with the same SNI settings.
- Making the new ServerOptionsSelectionCallback accessible in code.

JamesNK commented:
> Should there be a global SNI config? E.g. so you can have 10 endpoints all with the same SNI settings.

So make it available in EndpointDefaults?

Tratcher commented:
> > Should there be a global SNI config? E.g. so you can have 10 endpoints all with the same SNI settings.
> 
> So make it available in EndpointDefaults?

Right

halter73 committed:
SNI -> Sni

halter73 committed:
Add CertificateConfigLoader for easier testing

halter73 committed:
Add SniOptionsSelectorTests

halter73 committed:
Add more SniOptionsSelectorTests

- TODO: Add even more SniOptionsSelectorTests
  - Verify we clone all SslServerAuthenticationOptions properties
  - Verify every property from fallbackOptions are used correctly
- TODO: Integration tests

halter73 committed:
Address PR feedback

halter73 committed:
Support SNI config in EndpointDefaults

halter73 committed:
Test new internal UseHttps() overload

halter73 committed:
Add CloneSslOptionsClonesAllProperties test

Tratcher commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -277,26 +326,41 @@ public void Load()
+                    {
+                        // Ensure endpoint is reloaded if it used the default SNI config and it changed.
+                        // No need to configure httpsOptions for SNI since the SniOptionsSelector will now use the EndpointDefaults SNI config.
+                        endpoint.Sni = ConfigurationReader.EndpointDefaults.Sni;
```
What's the precedence order here? EndpointDefaults.Sni shouldn't take precedence over Endpoint.Certificate.

halter73 replied:
This is why I was apprehensive about adding EndpointDefaults.Sni to begin with. Endpoint.Certificate doesn't take precedence over Endpoint.Sni. Having precedence where Sni is prioritized over Certificate usually but not always is weird.

If any other property is defined in EndpointDefaults, but not directly on the Endpoint, we always treat it as if the property defined in EndpointDefaults was defined directly on the Endpoint. It doesn't makes sense to me to carve out an exception to this rule just for SNI.

If you want to use the Endpoint.Certificate don't define EndpointDefaults.Sni, or at least don't define it for a server name the endpoint serves and add an empty wildcard ("*") SNI so then it will fall back to the Endpoint.Certificate.

I would rather remove EndpointDefaults.Sni than have some special convoluted fallback logic just for EndpointDefaults.Sni.

halter73 replied:
`EndpointDefaults:Sni` is removed.

halter73 committed:
More tests

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/SniOptionsSelector.cs:
```diff
@@ -0,0 +1,223 @@
+                    if (serverName.EndsWith(suffix, StringComparison.OrdinalIgnoreCase))
+                    {
+                        sniOptions = options;
+                        break;
```
```suggestion
                        // _wildcardPrefixOptions is sorted with longest first, so we know the first match is the longest possible match.
                        break;
```

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/SniOptionsSelector.cs:
```diff
@@ -0,0 +1,223 @@
+            {
+                if (serverName is null)
+                {
+                    // There was no ALPN
```
```suggestion
                    // There was no SNI
```

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/SniOptionsSelector.cs:
```diff
@@ -0,0 +1,223 @@
+        // This won't catch issues related to mutable subproperties, but the existing subproperties look like they're mostly immutable.
+        // The exception are the ApplicationProtocols list which we clone and the ServerCertificate because of methods like Import() and Reset() :(
+        internal static SslServerAuthenticationOptions CloneSslOptions(SslServerAuthenticationOptions sslOptions) =>
+            new SslServerAuthenticationOptions
```
Should we ask the runtime to add a clone method on SslServerAuthenticationOptions?

halter73 replied:
Maybe. I added the test. I don't thing it's a huge deal either way.

Tratcher commented on src/Servers/Kestrel/Core/test/SniOptionsSelectorTests.cs:
```diff
@@ -0,0 +1,459 @@
+            Assert.Equal("WildcardPrefix", pathDictionary[aSubdomainOptions.ServerCertificate]);
+
+            // "*.example.org" is preferred over "*", but "*.example.org" doesn't match "example.org".
+            // REVIEW: Are we OK with "example.org" matching "*" instead of "*.example.org"? It feels annoying to me to have to configure example.org twice.
```
Did you run this past Barry?

The rule in similar wildcard systems is that `foo` does not match `*.foo`, it only matches `*`.

halter73 replied:
@blowdart 😄 

Not until now. I've written it to have the behavior you wanted where "\*.example.org" doesn't match "example.org" as annoying as it may be, because as I mention on the following line, the alternative would have "a.example.org" match "\*.a.example.org" before "\*.example.org", and that just seems wrong.

blowdart replied:
So *.example.org should not match example.org, but it should match a.example.org, b.example.org, a.b.example.org and so on. That's how SANs in certificates work as well (almost) so it's not surprising.

A wildcard of just * should match everything IMO.

So what chris said :)

Tratcher commented on src/Servers/Kestrel/Core/test/SniOptionsSelectorTests.cs:
```diff
@@ -0,0 +1,459 @@
+        }
+
+        [Fact]
+        public void WildcardOnlyMatchesNullServerNameDueToNoAlpn()
```
```suggestion
        public void WildcardOnlyMatchesNullServerNameDueToNoSni()
```

Tratcher reviewed (commented):
This looks really good. What's left on the draft?

halter73 committed:
Add functional test

halter73 commented on src/Servers/Kestrel/Core/src/Internal/SniOptionsSelector.cs:
```diff
@@ -0,0 +1,223 @@
+                }
+
+                // SkipValidation is only true in tests.
+                if (!certifcateConfigLoader.SkipValidation && sslOptions.ServerCertificate is X509Certificate2 cert2)
```
I want a unique cert instance per SNI section, and I was concerned about the performance implications of loading real ones so many times.

Tratcher replied:
? This check only verifies EKUs, the cert is already loaded.

halter73 replied:
But I want a unique cert instance per SNI section. Are you sure that creating a new X509Certificate2 for an already-loaded cert is effectively free?

javiercn commented on src/Servers/Kestrel/Core/src/Internal/Certificates/ICertificateConfigLoader.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Certificates
+{
+    internal interface ICertificateConfigLoader
```
Is this for testing/mocking? 

You could consider using an Abstract Base Class instead (channeling my inner rynowak here) the interface doesn't buy you a lot here since I don't think you plan to have multiple implementations.

halter73 replied:
It's internal and just for testing. Is there some reason using an abstract base class is better for testing?

halter73 commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -329,12 +345,22 @@ public void Load()
+                    if (endpoint.SNI.Count == 0)
                     {
-                        throw new InvalidOperationException(CoreStrings.NoCertSpecifiedNoDevelopmentCertificateFound);
+                        if (httpsOptions.ServerCertificate == null && httpsOptions.ServerCertificateSelector == null)
```
I'm not sure if this comment was originally made for the non-SNI branch which this is, but it's loaded by the `Options.ApplyDefaultCert(httpsOptions)` line above like before. In the SNI branch, the dev cert will be used ass a fallback if necessary.

halter73 commented on src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:
```diff
@@ -282,6 +196,140 @@ await using (sslDuplexPipe)
+            }
+
+            // REVIEW: Should we write any event before this?
+            KestrelEventSource.Log.TlsHandshakeStart(context, sslOptions);
```
There are two branches where we start the handshake. Since we don't have the `sslOptions` when we start the callback-based version of AuthenticateAsServerAsync, we log it in the callback itself. I now have a comment in the old HttpsConnectionMiddlware about how to completely consolidate these two branches (we already share methods where possible). I don't want to do this until https://github.com/dotnet/runtime/issues/40402 is resolved though, since it would break a lot of our client cert logic.

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs:
```diff
@@ -73,14 +87,25 @@ private IEnumerable<EndpointConfig> ReadEndpoints()
-                //    "ClientCertificateMode" : "NoCertificate"
+                //     "Url": "https://*:5463",
+                //     "Protocols": "Http1AndHttp2",
+                //     "SslProtocols": [ "Tls11", "Tls12", "Tls13"],
+                //     "Certificate": {
+                //         "Path": "testCert.pfx",
+                //         "Password": "testPassword"
+                //     },
+                //     "ClientCertificateMode" : "NoCertificate",
+                //     "Sni": {
```
Offline discussion:
Throw if any of these SSL settings are enabled for an HTTP endpoint.

halter73 committed:
SkipValidation -> IsTestMock

halter73 committed:
Remove EndpointDefaults.Sni

halter73 committed:
Validate that HTTPS config is not applied to non-HTTPS endpoints

halter73 committed:
Fix CloneSslOptionsClonesAllProperties test on Ubuntu 16.04

Tratcher commented on src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs:
```diff
@@ -160,6 +165,37 @@ void ConfigureListenOptions(ListenOptions listenOptions)
             }
         }
 
+        [Fact(Skip = "https://github.com/dotnet/runtime/issues/40402")]
```
You could mark this as Quarantined and then we'll unquarantine it once it starts passing.

halter73 committed:
Address PR feedback

- Hopefully this is it!
